### PR TITLE
Access list to share manager

### PR DESCRIPTION
**Pull request by rullzer**
Revived this PR to get rid of the old code

What we do now is offload the accesslist creation to the share providers. Which allows for way more efficient queries.

The manager function now can handle recursion and directly filter out users the currently have no access to the Node (delete group shares for example).

The helper class should be able to assist for notifications.

@nickvergessen @schiessle lets see if we can get this moving!

TODO: 
- [X] add  accesslist function to circles provider
  * Issue at https://github.com/nextcloud/circles/issues/46

### CODE REVIEW

File: apps/sharebymail/lib/ShareByMailProvider.php [hunk 1]
```diff
@@ -767,4 +767,8 @@ public function getSharesInFolder($userId, Folder $node, $reshares) {
 		return $shares;
 	}
 
+	public function getAccessList($nodes, $currentAccess) {
+		return ['users' => [], 'remote' => false, 'public' => false];
```
Comment (schiessle) [inline apps/sharebymail/lib/ShareByMailProvider.php]: Shouldn't we check here if the file or a file in the folder is shared by mail and set 'public' to true?

Comment (rullzer) [inline apps/sharebymail/lib/ShareByMailProvider.php]: Yeah didn't think about it yet. Just had to implement something that did not break. I'm fine with both so whatever you prefer.

Comment (rullzer) [inline apps/sharebymail/lib/ShareByMailProvider.php]: Ok added

**Comment by rullzer** [comment]
@nickvergessen @schiessle please have a look

File: lib/private/Encryption/File.php [hunk 2]
```diff
@@ -63,26 +76,34 @@ public function getAccessList($path) {
...
 		$public = $resultForParents['public'] || $resultForParents['remote'];
 
 
 		// Find out who, if anyone, is sharing the file
-		$resultForFile = \OCP\Share::getUsersSharingFile($ownerPath, $owner, false, false, false);
-		$userIds = \array_merge($userIds, $resultForFile['users']);
-		$public = $resultForFile['public'] || $resultForFile['remote'] || $public;
+		if ($file !== null) {
+			$resultForFile = $this->shareManager->getAccessList($file, false);
+			$userIds = \array_merge($userIds, $resultForFile['users']);
```
Comment (nickvergessen) [inline lib/private/Encryption/File.php]: remove leading slash?

File: lib/public/Share/IManager.php [hunk 3]
```diff
@@ -192,6 +192,36 @@ public function groupDeleted($gid);
...
+	 * ]
+	 *
+	 * This is required for encryption/activity
+	 *
+	 * @param \OCP\Files\Node $path
+	 * @param bool $recursive Should we check all parent folders as well
+	 * @return array
+	 * @since 9.2.0
+	 */
+	public function getAccessList(\OCP\Files\Node $path, $recursive = true);
```
Comment (nickvergessen) [inline lib/public/Share/IManager.php]: Missing `, $currentAccess = false` ?

File: lib/private/Share20/ShareHelper.php [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+	}
+
+	/**
+	 * If a user has access to a file
+	 *
+	 * @param Node $node
+	 * @param array $users Array of userIds
+	 * @return array Mapping $uid to an array of nodes
+	 */
+	public function getPathsForAccessList(Node $node, $users) {
```
Comment (nickvergessen) [inline lib/private/Share20/ShareHelper.php]: No public API?

**Review by MorrisJobke — DISMISSED** [review]
@rullzer See comments by @nickvergessen

**Comment by rullzer** [comment]
@nickvergessen I have not yet found a more efficient and correct way to fetch all paths. Basically if we want to path to be correct then it takes time because we need to setup the FS.

**Comment by rullzer** [comment]
Had a quick look. @nickvergessen your edits make sense to me. But we'll have to do some encryption tests. And maybe add more tests as well.

**Comment by nickvergessen** [comment]
You are going to do that @rullzer or "someone"

**Comment by rullzer** [comment]
I'm not doing it now or any time soon I'm afraid...

**Comment by nickvergessen** [comment]
Fixed some tests, open tasks:
- [ ] Tests for ShareHelper
- [ ] Inconsistent result from DefaultShareProvider with currentAccess = false

**Comment by rullzer** [comment]
I just checked and encryption still works here!

**Comment by rullzer** [comment]
What I tested:

1. setup encrypted instance
2. have 3 users (with valid keys): user0, user1, user2
3. user0 has a file `foo/bar/baz.txt`
4. user0 shares `foo` with user1
5. user0 shares `baz.txt` with user2
6. verify everybody can see the content of baz.txt

Now each user modifies the file and verified that others can see the change.

**Comment by rullzer** [comment]
Ah mmm no I see a bug I think. Letr me think how to fix this properly

**Comment by rullzer** [comment]
:+1: for me now. Seems to do all the tricks!

@schiessle can you have a look as well?

**Comment by schiessle** [comment]
'public' need to be true as soon as one remote-share, email-share or public link exists. I don't see that this is the case here, but maybe I missed something.

**Comment by rullzer** [comment]
don't we ave remote for remote?

**Comment by rullzer** [comment]
@schiessle also see https://github.com/nextcloud/server/pull/2834#discussion_r93749606 ;)

**Comment by schiessle** [comment]
Yes you are right, remote shares are separate, only the encryption code merges them: https://github.com/nextcloud/server/blob/master/lib/private/Encryption/File.php#L79

But mail-shares and public links where handled by the share api as the same: https://github.com/nextcloud/server/blob/master/lib/private/Share/Share.php#L269

So we either need to keep this behaviour for mail shares and public links or extend the logic on the encryption code to merge public, remote and mail.

**Comment by rullzer** [comment]
Let me just replace the public key. Because the code doesn't care so that seems the easy way.

**Comment by schiessle** [comment]
> also see #2834 (comment) ;)

Yes, makes probably sense to show mail shares as mail shares but then we need to update it here https://github.com/nextcloud/server/blob/master/lib/private/Encryption/File.php#L79 and here https://github.com/nextcloud/server/blob/master/lib/private/Encryption/File.php#L85

**Comment by rullzer** [comment]
@schiessle I just made it public now. As that keep the behavior as is for now.

**Comment by rullzer** [comment]
CI is happy (well 1 timed out).
Also activities and encryption seem to work properly for me still!

@schiessle @nickvergessen what do you say? Lets get this in!

**Comment by nickvergessen** [comment]
:+1:

**Review by LukasReschke — APPROVED** [review]

**Comment by daita** [comment]
Circles complies with the new ShareProvider

**Pull request merged**